File: bio1/week2/week2.go
package week2

import (
	"bioinformatics/bio1/week1"
	"fmt"
)

func CalculateSkew(genome string) string {
	var resultArr []int
	resultArr = append(resultArr, 0)
	for i, s := range genome {
		if s == 'G' {
			resultArr = append(resultArr, resultArr[i]+1)
			continue
		} else if s == 'C' {
			resultArr = append(resultArr, resultArr[i]-1)
			continue
		}

		resultArr = append(resultArr, resultArr[i])
	}

	retVal := fmt.Sprint(resultArr)
	return retVal
}

func CalculateLowestPointIndexes(genome string) []int {
	var scoreArr []int
	score := 0
	scoreArr = append(scoreArr, score)
	min := int(^uint(0) >> 1)
	for i, s := range genome {
		if s == 'G' {
			score = scoreArr[i] + 1
		} else if s == 'C' {
			score = scoreArr[i] - 1
		} else {
			score = scoreArr[i]
		}
		if score < min {
			min = score
		}
		scoreArr = append(scoreArr, score)
	}

	var resultArr []int
	for i, v := range scoreArr {
		if v <= min {
			resultArr = append(resultArr, i)
		}
	}
	return resultArr
}

func HammingDistance(kmer1, kmer2 string) int {
	counter := 0
	for i, _ := range kmer1 {
		if kmer1[i] != kmer2[i] {
			counter++
		}
	}
	return counter
}

func ApproximatePatternMatching(pattern, genome string, tolerance int) []int {
	var resultArr []int
	k := len(pattern)
	genomeLen := len(genome)
	for i, _ := range genome {
		if i+k > genomeLen {
			break
		}

		comparedKmer := genome[i : i+k]
		if HammingDistance(pattern, comparedKmer) <= tolerance {
			resultArr = append(resultArr, i)
		}
	}
	return resultArr
}

func ApproximatePatternMatchingCount(pattern, genome string, tolerance int) int {
	return len(ApproximatePatternMatching(pattern, genome, tolerance))
}

func FreqWordWithApproximatePatternMatching(genome string, k, tolerance int) []string {
	setMap := make(map[string]struct{})
	finalMap := make(map[string]int)
	var resultArr []string
	genomeLen := len(genome)
	max := 0
	for i, _ := range genome {
		if i+k > genomeLen {
			break
		}
		kmer := genome[i : i+k]
		setMap = mutate(kmer, tolerance)

		//join to finalMap
		for setMapKey, _ := range setMap {
			result := finalMap[setMapKey] + 1
			finalMap[setMapKey] = result
			if max < result {
				max = result
			}
		}
	}

	for k, v := range finalMap {
		if v >= max {
			resultArr = append(resultArr, k)
		}
	}

	return resultArr
}

func FreqWordWithApproximatePatternMatchingAndReverse(genome string, k, tolerance int) (string, int) {
	finalMap := make(map[string]int)
	// finalReverseMap := make(map[string]int)
	result := ""
	genomeLen := len(genome)
	max := 0
	for i, _ := range genome {
		if i > genomeLen-k {
			break
		}
		kmer := genome[i : i+k]
		finalMap[kmer]++
	}

	for key, _ := range finalMap {
		kReverse := week1.ReverseComplement(key)
		if key == kReverse {
			continue
		}
		count := finalMap[key] + finalMap[kReverse]

		mutationMap := mutate(key, tolerance)
		for m, _ := range mutationMap {
			mReverse := week1.ReverseComplement(m)
			count += finalMap[m] + finalMap[mReverse]
		}

		if count > max {
			result = key + " " + kReverse
			max = count
		}
	}

	return result, max
}

func reverse(s string) string {
	chars := []rune(s)
	for i, j := 0, len(chars)-1; i < j; i, j = i+1, j-1 {
		chars[i], chars[j] = chars[j], chars[i]
	}
	return string(chars)
}

var base = [4]string{"A", "T", "G", "C"}

func mutate(kmer string, tolerance int) map[string]struct{} {
	setMap := make(map[string]struct{})
	if tolerance < 1 {
		setMap[kmer] = struct{}{}
		return setMap
	}
	setMap = mutateOne(kmer)
	for i := 0; i < tolerance-1; i++ {
		localMap := make(map[string]struct{})
		for k, _ := range setMap {
			mutationMap := mutateOne(k)

			//join to localMap
			for mutKey, _ := range mutationMap {
				if _, ok := localMap[mutKey]; !ok {
					localMap[mutKey] = struct{}{}
				}
			}
			delete(setMap, k)
		}
		//join to setMap
		for localKey, _ := range localMap {
			if _, ok := setMap[localKey]; !ok {
				setMap[localKey] = struct{}{}
			}
		}
	}
	return setMap
}

func mutateOne(kmer string) map[string]struct{} {
	setMap := make(map[string]struct{})
	for i := range kmer {
		for _, b := range base {
			val := kmer[:i] + b + kmer[i+1:]
			if val == kmer {
				continue
			}
			setMap[val] = struct{}{}
		}
	}
	return setMap
}

File: bio1/week2/week2.py
def approximate_pattern_matching(pattern, genome, tolerance):
	resultArr = []
	k = len(pattern)
	genomeLen = len(genome)
	for i in range(genomeLen - k + 1):
		comparedKmer = genome[i : i+k]
		if(hamming_distance(pattern, comparedKmer) <= tolerance):
			resultArr.append(i)
	return resultArr

def hamming_distance(kmer1, kmer2):
    counter = 0
    kmer1Len = len(kmer1)
    for i in range(kmer1Len):
        if(kmer1[i] != kmer2[i]):
            counter += 1
    return counter

def immediate_neighbors(pattern):
	"""
	https://stepik.org/lesson/3014/step/1?unit=8228
	"""
	nucleotides = ['A', 'T', 'G', 'C']
	neighborhoods = set()
	for i,v in enumerate(pattern):
		for _,n in enumerate(nucleotides):
			if(n == v):
				continue

			neighbor = pattern[:i] + n + pattern[i+1:]
			neighborhoods.add(neighbor)
	
	return neighborhoods

def neighbors(pattern, d):
	"""
	https://stepik.org/lesson/3014/step/3?unit=8228
	"""

	if d == 0:
		return set(pattern)
	
	nucleotides = {'A', 'T', 'G', 'C'}
	if len(pattern) == 1:
		return nucleotides	
	
	neighborhood = set()
	suffix_neighbors = neighbors(pattern[1:], d)

	for i,val in enumerate(suffix_neighbors):
		if hamming_distance(pattern[1:], val) < d:
			for _,n in enumerate(nucleotides):
				neighborhood.add(n + val)
			else:
				neighborhood.add(pattern[0] + val)
	
	return neighborhood

results = neighbors('CGAGACTTACTC', 3)
for k,v in enumerate(results):
	with open("test.txt", "a") as myfile:
		myfile.write(v + ' ')
	